File: 14/code.ts
import {
  decodeText,
  chunkLines,
  filter,
  takeWhile,
  map,
  mapNonNullable,
  toArray,
} from '../iter.ts';

const stdin = chunkLines(decodeText(Deno.iter(Deno.stdin)));

const templates = await toArray(
  takeWhile(stdin, (line) => line),
);

let template = templates[0];

console.log('template', template);

const rules = new Map();
for await (const line of stdin) {
  const components = line.trim().split(' -> ');
  rules.set(components[0], components[1]);
}

console.log('rules', rules);

for (const i of range(0, 10)) {
  let newTemplate = '';
  for (const j of range(0, template.length - 1)) {
    const pair = template.slice(j, j + 2);
    const insertion = rules.get(pair);

    if (insertion) {
      newTemplate += pair[0] + insertion;
    } else {
      newTemplate += pair[0];
    }
  }
  template = newTemplate + template[template.length - 1];
}

//console.log('done', template);

let letterCounts = new Map();
for (const letter of template) {
  letterCounts.set(letter, (letterCounts.get(letter) ?? 0) + 1);
}

console.log('counts', letterCounts);

const [min, max] = [...letterCounts.values()].reduce(([min, max], count) => [
  Math.min(count, min),
  Math.max(count, max),
], [Infinity, 0]);

console.log('final value', max - min);

function *range(from: number, to: number) {
  for (let i = from; i < to; i++) {
    yield i;
  }
}

File: 14/code.rs
use itertools::Itertools;
use std::collections::HashMap;
use utils;

fn main() {
    let mut lines = utils::read_input_file_lines();

    let template = lines.next().expect("must have a template");
    lines.next();

    let mut pair_counts: HashMap<(char, char), usize> = template.chars().tuple_windows().counts();

    let rules: HashMap<(char, char), char> = lines
        .map(|line| {
            let (pair, result) = line.split_once(" -> ").expect("should split");
            (
                pair.chars().next_tuple().expect("invalid pair"),
                result.chars().take(1).next().unwrap(),
            )
        })
        .collect();

    for _ in 0..10 {
        let mut new_pair_counts = HashMap::new();
        for ((c1, c2), count) in pair_counts {
            if let Some(insertion) = rules.get(&(c1, c2)) {
                *new_pair_counts.entry((c1, *insertion)).or_insert(0) += count;
                *new_pair_counts.entry((*insertion, c2)).or_insert(0) += count;
            }
        }
        pair_counts = new_pair_counts;
    }

    let mut char_counts: HashMap<char, usize> =
        HashMap::from([(template.chars().next().expect("first char"), 1)]);
    for ((_c1, c2), count) in &pair_counts {
        *char_counts.entry(*c2).or_insert(0) += count;
    }

    match char_counts.values().minmax() {
        itertools::MinMaxResult::MinMax(min, max) => {
            println!("final value {}", max - min);
        }
        _ => {
            println!("failed to get a min max");
        }
    }
}
